aten/src/ATen/native/native_functions.yaml

```diff
@@ -2192,9 +2192,12 @@
   device_check: NoCheck
   device_guard: False
 
+# decomposes to eye.m
 - func: eye(int n, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
 - func: eye.m(int n, int m, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+  dispatch:
+    CompositeExplicitAutograd: eye
```

Comment:
Why is this `CompositeExplicit`? I think it doesn't have a derivative, does it?

Reply:
So what's actually going on here is that CompositeImplicitAutograd get decomposed automatically for purposes of Python key tracing after https://github.com/pytorch/pytorch/pull/81984 but CompositeExplicitAutograd does not. But eye.m decomposes into an empty alloc and then an out= call, and I don't want that to happen, I want it to get traced as a single unit. Adding this line forces this behavior.

Reply:
But wouldn't this have some secondary effects like weirder errors of the form "`eye` does not have a derivative" or so if you try to differentiate through it? Or is it fine because it always returns a tensor with `requires_grad=False`? What about forward AD? CC @albanD

Reply:
Actually that is not true, you can do `torch.eye(2, requires_grad=True)`. But the autograd is already handled at a higher level and in a special case for factory functions.

@ezyang I'm not sure to understand why you need to change that though? This value is ignored for factory functions anyways?

Reply:
- In https://github.com/pytorch/pytorch/pull/81984, we make the Python key decompose on CompositeImplicitAutograd.
- The side effect of this is that at::eye, which is CompositeImplicitAutograd, gets decomposed into at::eye_out
- at::eye_out is problematic because it out-variants in general are not "composite compliant" - e.g. doesn't work for all Tensor Subclasses (vmap, autograd). Technically this usage of eye_out is acceptable but our composite compliance testing catches usages of out= variants in composite operations.
- So the resolution is to make eye CompositeExplicitAutograd, like we have done to e.g. at::empty_like

Reply:
Note that for `empty_like`, when we did so, we added the following entry in `derivatives.yaml`:
```yaml
- name: empty_like(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
  self: non_differentiable
  output_differentiability: [False]
```
Should we do the same for `eye` then?

Reply:
I think it is not necessary as there are no tensor inputs